resilience4j-ratpack/src/main/java/io/github/resilience4j/ratpack/bulkhead/BulkheadMethodInterceptor.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2017 Jan Sykora
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.github.resilience4j.ratpack.bulkhead;
+
+import com.google.inject.Inject;
+import io.github.resilience4j.bulkhead.BulkheadFullException;
+import io.github.resilience4j.bulkhead.BulkheadRegistry;
+import io.github.resilience4j.bulkhead.operator.BulkheadOperator;
+import io.github.resilience4j.ratpack.recovery.RecoveryFunction;
+import io.reactivex.Flowable;
+import io.reactivex.Observable;
+import io.reactivex.Single;
+import org.aopalliance.intercept.MethodInterceptor;
+import org.aopalliance.intercept.MethodInvocation;
+import ratpack.exec.Promise;
+import ratpack.util.Exceptions;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * A {@link MethodInterceptor} to handle all methods annotated with {@link Bulkhead}. It will
+ * handle methods that return a Promise only. It will add a transform to the promise with the bulkhead and
+ * fallback found in the annotation.
+ */
+public class BulkheadMethodInterceptor implements MethodInterceptor {
+
+    @Inject(optional = true)
+    private BulkheadRegistry registry;
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public Object invoke(MethodInvocation invocation) throws Throwable {
+        Bulkhead annotation = invocation.getMethod().getAnnotation(Bulkhead.class);
```

Comment:
I know that currently all interceptor in ratpack module are implemented this way, but maybe we can find better solution.

Reply:
Let's make this refactoring independent from this PR.